sympy/simplify/cse_main.py

```diff
@@ -40,6 +40,197 @@
+        subs_mapping : ``Dict[Symbol, Expr]``
+            Pairs of substitutions present in ``exprs``.
+        as_list : ``bool``, default is ``True``
+            Whether the argument passed to ``exprs`` should be converted to a
+            ``list``. With the default of ``True``, wif an iterable is passed to
+            ``exprs`` then it is converted to a ``list``, or if a single
+            ``Expr`` is passed then it is wrapped into a single-item list. If
+            ``False``, then the argument passed to ``exprs`` is left as is,
+            meaning that the value returned by the ``exprs`` property is not
+            necessarily a ``list``. This parameter is similar to the ``list``
+            parameter of the ``cse`` function.
+
+        """
+
+        # Handle the case if just one expression was passed.
+        if as_list and isinstance(exprs, (Basic, MatrixBase)):
+            exprs = [exprs]
+
+        # Ensure items in argument passed to ``exprs`` are SymPy expressions
+        if not isinstance(exprs, (Basic, MatrixBase)):
+            for i, expr in enumerate(exprs):
+                if not isinstance(expr, (Basic, MatrixBase)):
+                    msg = (
+                        f'Expression {expr} passed at index {i} must be a '
+                        f'SymPy expression, not a {type(expr)}'
+                    )
+                    raise TypeError(msg)
+
+        # Ensure argument passed to ``subs_mapping`` is a ``dict``
+        if not isinstance(subs_mapping, dict):
+            msg = (
+                f'`subs_mapping` must be a dictionary mapping SymPy symbols '
+                f'to SymPy expressions, not a {type(subs_mapping)}'
+            )
+            raise TypeError(msg)
+
+        # Ensure symbol-expression pairs in ``subs_mapping`` are correct types
+        for i, (symb, expr) in enumerate(subs_mapping.items()):
+            if not isinstance(symb, Symbol):
+                msg = (
+                    f'Key {symb} at index {i} in `subs_mapping` must be a '
+                    f'SymPy symbol, not a {type(symb)}'
+                )
+                raise TypeError(msg)
+            if not isinstance(expr, (Basic, MatrixBase)):
+                msg = (
+                    f'Value {expr} at index {i} in `subs_mapping` must be a '
+                    f'SymPy expression, not a {type(expr)}'
+                )
+                raise TypeError(msg)
+
+        return cls((reps_toposort(list(subs_mapping.items())), exprs))
+
+    def as_tuple(self):
+        """Convert the ``CseExpr`` instance to a ``tuple``.
+
+        This method exists purely for backwards compatibility where ``cse``
+        previously returned a ``tuple``. While this class is fully backwards
+        compatible with the previous return type, the method exists to allow
+        the easy conversion to a pure ``tuple`` if that's required by a user.
+
+        """
+        return tuple(self)
+
+    @property
+    def subs(self):
```

Comment:
I think we should call this something else (maybe `replacements`). `subs` is usually a method in SymPy.

Reply:
Sticking with `replacements` and `reduced_exprs` (one might think that `exprs` are the expressions before `cse`) makes sense to me since that is what it is currently called in the `cse` function.

Reply:
Names changed to `replacements` and `reduced_exprs` in https://github.com/sympy/sympy/pull/24649/commits/b101136726ace893c1d41a303b2462d724a95636.